File: scratch/extracted/graph-bfs/optimize.cpp
class Solution {
public:
  vector < int > bfs(vector<vector<int>> & graph, int start) {
  // Initialize an unordered set to keep track of visited nodes.
  unordered_set < int > visited;
  // Initialize a queue for BFS traversal.
  queue < int > q;
  // Initialize a vector to store the result (BFS traversal order).
  vector < int > result;

  // Mark the starting node as visited and enqueue it.
  visited.insert(start);
  q.push(start);

  // Perform BFS until the queue is empty.
  while (!q.empty()) {
    // Dequeue a node from the front of the queue.
    int node = q.front();
    q.pop();
    // Add the dequeued node to the result vector.
    result.push_back(node);

    // Iterate through the neighbors of the current node.
    for (int neighbor : graph[node]) {
           // If the neighbor has not been visited yet...
      if (visited.find(neighbor) == visited.end()) {
        // Mark the neighbor as visited.
        visited.insert(neighbor);
        // Enqueue the neighbor for further exploration.
        q.push(neighbor);
      }
    }
  }

  // Return the BFS traversal result.
  return result;
}
};

File: scratch/extracted/graph-bfs/optimize.py
# Definition for a graph node.
# class GraphNode:
#     def __init__(self, val=0, neighbors=None):
#         self.val = val
#         self.neighbors = neighbors if neighbors is not None else []
def bfs(graph, start):
    # visited: set to keep track of visited nodes to prevent cycles and redundant processing.
    visited = set()
    # queue: queue for BFS traversal, ensuring breadth-wise exploration.
    queue = [start]
    # result: list to store the BFS traversal order, maintaining the order of visited nodes.
    result = []

    # Mark the starting node as visited to avoid revisiting it.
    visited.add(start)

    # While the queue is not empty, continue exploring the graph.
    while queue:
        # Dequeue a vertex from queue, retrieving the next node to process.
        node = queue.pop(0)
        # Add the dequeued vertex to the result list, recording the traversal order.
        result.append(node)

        # Get all adjacent vertices of the dequeued vertex node.
        # If a adjacent has not been visited, then mark it visited
        # and enqueue it for further exploration.
        for neighbor in graph[node]:
            # Check if the neighbor has already been visited.
            if neighbor not in visited:
                # Mark the neighbor as visited to prevent cycles.
                visited.add(neighbor)
                # Enqueue the neighbor for later processing.
                queue.append(neighbor)

    # Return the result list containing the BFS traversal order.
    return result
